packages/api/src/hooks/useCreateActivityStatusRenderer.tsx

```diff
@@ -1,104 +1,107 @@
-import type { WebChatActivity } from 'botframework-webchat-core';
+import React, { memo, useMemo } from 'react';
 
-import useGetSendTimeoutForActivity from './useGetSendTimeoutForActivity';
-import useTimePassed from './internal/useTimePassed';
+import useGetKeyByActivity from '../hooks/useGetKeyByActivity';
+import useSendStatusByActivityKey from '../hooks/useSendStatusByActivityKey';
 import useWebChatAPIContext from './internal/useWebChatAPIContext';
 
-const {
-  ActivityClientState: { SEND_FAILED, SENDING, SENT }
-} = Constants;
+import type { ReactNode, VFC } from 'react';
+import type { RenderActivityStatus } from '../types/ActivityStatusMiddleware';
+import type { SendStatus } from '../providers/ActivitySendStatus/SendStatus';
+import type { WebChatActivity } from 'botframework-webchat-core';
 
-type ActivityStatusContainerProps = {
+type InnerActivityStatusContainerProps = {
   activity: WebChatActivity;
   hideTimestamp: boolean;
   nextVisibleActivity: WebChatActivity;
+  sendStatus: SendStatus;
 };
 
-const ActivityStatusContainer: VFC<ActivityStatusContainerProps> = ({
-  activity,
-  hideTimestamp,
-  nextVisibleActivity
-}) => {
-  const { activityStatusRenderer: createActivityStatusRenderer } = useWebChatAPIContext();
-  const getSendTimeoutForActivity = useGetSendTimeoutForActivity();
-
-  // SEND_FAILED from the activity is ignored, and is instead based on styleOptions.sendTimeout.
-  // Note that the derived state is time-sensitive. The useTimePassed() hook is used to make sure it changes over time.
-  const {
-    from: { role }
-  }: WebChatActivity = activity;
-
-  const fromUser = role === 'user';
-  let activitySent: boolean;
-  let sendTimeoutAt: number;
-
-  if (fromUser) {
-    const state = activity.channelData?.state;
-    const sendTimeout = getSendTimeoutForActivity({ activity });
-
-    activitySent = state !== SENDING && state !== SEND_FAILED;
-
-    // If no timestamp, we assume the "sending" will be timed out as "send failed".
-    sendTimeoutAt = !activitySent
-      ? new Date(activity.localTimestamp || new Date(0).toISOString()).getTime() + sendTimeout
-      : 0;
-  } else {
-    activitySent = true;
-    sendTimeoutAt = 0;
+const ActivityStatusContainerCore: VFC<InnerActivityStatusContainerProps> = memo(
+  ({ activity, hideTimestamp, nextVisibleActivity, sendStatus }) => {
+    const { activityStatusRenderer: createActivityStatusRenderer }: { activityStatusRenderer: RenderActivityStatus } =
+      useWebChatAPIContext();
+
+    return createActivityStatusRenderer({
+      activity,
+      hideTimestamp,
+      nextVisibleActivity, // "nextVisibleActivity" is for backward compatibility, please remove this line on or after 2022-07-22.
+      sameTimestampGroup: hideTimestamp, // "sameTimestampGroup" is for backward compatibility, please remove this line on or after 2022-07-22.
+      sendState: sendStatus === 'send failed' || sendStatus === 'sent' ? sendStatus : 'sending'
```

Comment:
Is `sendStatus` typed? Should we use constants here?

```suggestion
      sendState: sendStatus === SEND_FAILED || sendStatus === SENT ? sendStatus : SENDING
```

Reply:
This is actually intentional. Feel free to comment:

- Let's stay away from the `Constants.SendState` thing we introduced before TypeScript
   - The constant can be replaced/typechecked by TypeScript without worry about typo
- I am unsure if we should go `enum ABC` or just `type ABC = 'one' | 'two' | 'three'`

Reply:
Fixed by using internal constant.

Reply:
I'd prefer `type ABC = 'one' | 'two' | 'three'` :)